orc8r/cloud/go/services/certifier/servicers/certifier.go

```diff
@@ -712,51 +714,95 @@ func (srv *CertifierServer) getPolicyDecisionFromTokenMany(tokens *certprotos.To
 	return &certprotos.GetPolicyDecisionResponse{Effect: finalEffect}, nil
 }
 
-func (srv *CertifierServer) getPolicyDecisionFromToken(token string, reqResource *certprotos.Resource) (certprotos.Effect, error) {
+func (srv *CertifierServer) getPolicyDecisionFromToken(ctx context.Context, token string, reqResource *certprotos.RequestResource) (certprotos.Effect, error) {
 	policy, err := srv.store.GetPolicy(token)
 	if err != nil {
 		return certprotos.Effect_UNKNOWN, status.Errorf(codes.Internal, "failed to get policy from db, checking next policy: %v", err)
 	}
 	effect := certprotos.Effect_UNKNOWN
+
+	// Networks are registered with tenants, hence any tenant scoped policies
+	// have additional policies for their networks.
+	tenantNetworkResource, err := getTenantPolicyNetworkResources(ctx, policy)
+	policy.Resources.Resources = append(policy.Resources.Resources, tenantNetworkResource...)
+
 	for _, policyResource := range policy.Resources.Resources {
 		actionEffect := checkAction(reqResource, policyResource)
-		if actionEffect == certprotos.Effect_DENY {
-			return certprotos.Effect_DENY, nil
-		}
 		resourceEffect := checkResource(reqResource, policyResource)
-		if resourceEffect == certprotos.Effect_DENY {
-			return certprotos.Effect_DENY, nil
-		}
 		// The effects from both action and resource should match
 		// for the effect of the policy to apply to the request.
 		if actionEffect != certprotos.Effect_UNKNOWN &&
 			resourceEffect != certprotos.Effect_UNKNOWN &&
 			actionEffect == resourceEffect {
 			effect = actionEffect
 		}
+
+		if actionEffect == certprotos.Effect_DENY &&
+			resourceEffect == certprotos.Effect_DENY {
+			return certprotos.Effect_DENY, nil
+		}
 	}
 
 	return effect, nil
 }
 
-// checkResource checks if the requested resource is authorized by the policy
-func checkResource(reqResource, policyResource *certprotos.Resource) certprotos.Effect {
-	// The policy does not handle this case, so return UNKNOWN.
-	if reqResource.ResourceType != policyResource.ResourceType {
-		return certprotos.Effect_UNKNOWN
+func getTenantPolicyNetworkResources(ctx context.Context, policy *certprotos.Policy) ([]*certprotos.PolicyResource, error) {
```

Comment:
I also thought about this... if had something like
```
{
	Effect:   certprotos.Effect_ALLOW,
	Action:   certprotos.Action_WRITE,
	Resource: &certprotos.PolicyResource_Tenant{Tenant: &certprotos.TenantResource{Tenant: "0"}},
},
{
	Effect:   certprotos.Effect_DENY,
	Action:   certprotos.Action_WRITE,
	Resource: &certprotos.PolicyResource_Tenant{Tenant: &certprotos.TenantResource{Tenant: "1"}},
},
```
Then we can't just lump them in one list of networks because we want to apply different effects on them. All of tenant 0's networks need to be granted write access and all of tenant 1's networks need to be denied write access. This is looping through all tenant policies and getting the networks for each

I'll split the inner for loop out for clarity though, maybe putting it all together is confusing